### ISSUE TITLE
Implement the data-parallel execution path behind the distributed planner

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#97]
> Filed against the `feature_request.yml` template. Category: **Optimizer / Loss Function / Autograd** (training runtime).

## Problem

The distributed planning layer is finished and executes nothing.

`docs/book/src/distributed.md`:

> The `distributed` feature provides typed meshes, placements, collective descriptors, and validation for data parallel, tensor parallel, pipeline, FSDP, and ZeRO plans. **It is a planning surface, not an end-to-end distributed training runtime.** [...] The current limitation is deliberate: there is no promise that a model can be trained across hosts.

`docs/plan/roadmap.md` records 16 completed `dist` rows with the note "Planning and placement layer only; there is no distributed execution path."

### What exists

Substantial, and it is not scaffolding. From `crates/incin-core/src/dist/mod.rs`'s public exports alone: `DeviceMesh`, `MeshSpec`, `ValidMesh`, `Placement`/`Sharded`/`Replicated`/`Partial`, `DataParallelPlan` and `DataParallelPlanBuilder`, `GradientDescriptor`, `CollectivePlan`, `CollectiveDescriptor`, `HybridPlanner`, `preflight`, `AgreedPlan`, `CommunicationEvidence`, plus `DistributedContext` with rendezvous, rank identity, and `TwoRankLaunchPlan`. `crates/incin-backends/src/dist/` adds a `CollectiveBackend` trait, a deterministic in-process reference transport, and a NCCL transport with buffers, wire format, and topology.

### What is missing

The step that connects a validated plan to a training loop. `crates/incin/src/train.rs:598` documents it:

> Returns [`TrainError::CollectivesUnavailable`] if the plan names more than one device.

That refusal is the honest behavior and it is also the entire gap. `DataParallelPlanBuilder` produces a `DataParallelPlan` full of `GradientDescriptor`s, `into_collective_plan()` turns it into a `CollectivePlan`, and nothing ever runs it.

Data-parallel is the right first execution path: every user wants it, it needs one collective pattern (gradient all-reduce), and it does not require the model to be partitioned.

## Proposed API sketch

The target is that the existing builder stops being a dead end:

```rust
use incin::prelude::*;
use incin_core::dist::{DeviceMesh, MeshSpec, Data};
use incin_core::typenum::U2;

// The mesh is already type-level and already proves its own validity.
type Mesh = MeshSpec<Data<U2>>;                       // 2-way data parallel, exists today

let plan = Trainer::plan()
    .devices(DeviceSet::cuda([0, 1])?)
    .mesh::<Mesh>()                                    // proposed
    .epochs(3)
    .build()?;

// Today: TrainError::CollectivesUnavailable { devices: 2 }
// Proposed: a real two-rank run.
let outcome = Trainer::new(plan).fit(&mut model, &mut optimizer, &mut loader, |out, batch| {
    out.cross_entropy_loss(&batch.targets)
})?;
```

Gradient synchronization is bucketed and overlapped rather than one collective per parameter:

```rust
// proposed - internal, sketched
// Backward already walks parameters in reverse topological order, so a bucket
// fills as soon as its last gradient is produced. Launch the all-reduce then,
// not after the whole backward pass.
let buckets = GradientBuckets::new(&plan, BucketPolicy::Bytes(25 * 1024 * 1024));
for ready in backward.stream_ready_gradients() {
    if let Some(bucket) = buckets.push(ready) {
        transport.all_reduce_async(bucket, ReduceOp::Mean, StreamId::Gradient)?;
    }
}
transport.wait_all(StreamId::Gradient)?;
```

## Shape and type-system invariants

The proofs already exist; this issue's job is to *lower from them* rather than re-derive anything at runtime.

| Invariant | Where it is already proven |
|---|---|
| Mesh topology is valid (nonzero degrees, product countable) | `ValidMesh`, compile time, with `#[diagnostic::on_unimplemented]` |
| World size matches the device count | `M: ValidMesh<World = U2>` bound |
| Gradient dtype is reducible | `DataParallelDType: CollectiveReductionDType<Mean>` |
| Collective compatibility across ranks | `preflight` and `AgreedPlan` |
| Every rank agrees on the plan | `CollectivePlan` digest and `TopologyFingerprint` |

The runtime's obligation is the complement: prove the requested physical devices, memory, transport, and topology actually exist, which `preflight` and `DeviceIdentity` already model. **A plan that cannot be executed correctly must refuse, not approximate.** That is the existing contract and this must not weaken it.

## Background

- **Bucketing is not an optimization, it is the design.** One all-reduce per parameter on a model with hundreds of parameters is dominated by launch latency. Bucketing by bytes and overlapping with the remainder of the backward pass is what makes data parallelism scale at all. A correct-but-serial implementation is the milestone floor, not the target.
- **The reduction is a mean, not a sum.** `DataParallelDType: CollectiveReductionDType<Mean>` already encodes this. Summing gradients across ranks silently scales the effective learning rate by the world size.
- **Optimizer state must stay consistent across ranks.** With identical initial state and identical averaged gradients it stays consistent by construction. Divergence means something upstream is wrong, so it should be *detectable* rather than assumed: a periodic state digest comparison is cheap.
- **Fail-stop over hang.** A rank dying mid-collective leaves the others blocked in a transport call. The preview already prefers fail-stop to a wrong answer; that has to extend to a recorded fail-stop rather than an indefinite wait.
- **`distributed-reference` is what makes this testable without hardware.** The deterministic in-process transport means the whole path can be covered in `ci.yml`, with NCCL adding only the transport-specific coverage in `hardware.yml`.

## How this interoperates

- **#98 is a hard prerequisite.** `incin-data` has no rank concept, so without it every rank draws identical batches and the run trains on one Nth of the data at N times the cost. Correct collectives over duplicated data is still wrong.
- **#82** fixes the `dist2-network` job, which currently is not disabled when no CUDA runner is registered and would schedule onto a GPU-less `ubuntu-latest`.
- **#99** builds on the collective plumbing this establishes, for FSDP, tensor, and pipeline parallelism.
- **`crates/incin-backends/tests/nccl_two_rank.rs`** already exists and is the natural home for the hardware-side proof.
- **`incin-telemetry` and `incin-viz`** should see per-rank step timing, otherwise a straggler rank is invisible.
- **Checkpointing:** single-node data parallelism replicates the model, so rank 0 saves and the existing typed snapshot contract is unchanged. Sharded checkpoints only become necessary with #99's FSDP, which is why #12 and #13 are scoped there rather than here.

## Acceptance criteria

- A model trains data-parallel across two devices and reaches the same loss trajectory as the single-device run within documented tolerance.
- `TrainError::CollectivesUnavailable` is returned only when collectives really are unavailable, not for every multi-device plan.
- The reference transport runs the same test in-process in `ci.yml`, so the path has coverage without hardware.
- The two-rank NCCL test runs in the `dist2-network` job once #82 has fixed its runner resolution.
- Gradients are bucketed and the reduction overlaps the backward pass, demonstrated by a measurement rather than asserted.
- A rank failing mid-step produces a recorded fail-stop within a bounded time, verified by a test that kills one.
- `docs/book/src/distributed.md` no longer describes data parallelism as planning-only.

## Alternatives

- **Gradient accumulation on one device.** Same effective batch size, no collectives, no second device. Worth documenting as the answer for users who do not need this; it does not reduce step time, which is the reason data parallelism exists.
- **One collective per parameter, no bucketing.** Simpler and measurably slower. Acceptable as a first landing only if the benchmark that shows why is landed with it.
- **Synchronous all-reduce after the full backward pass.** Simplest correct version; forfeits the overlap that is most of the benefit.

## Additional context

Relevant code: `crates/incin-core/src/dist/data_parallel.rs` (`DataParallelPlan`, `DataParallelPlanBuilder`, `GradientDescriptor`), `crates/incin-core/src/dist/plan/` (`CollectivePlan`, `preflight`, `AgreedPlan`), `crates/incin-core/src/dist/mesh.rs` (`MeshSpec`, `ValidMesh`), `crates/incin-core/src/dist/context/` (rendezvous, rank identity), `crates/incin-backends/src/dist/collective.rs` (`CollectiveBackend`), `crates/incin-backends/src/dist/reference.rs`, `crates/incin-backends/src/dist/nccl/transport.rs`, `crates/incin-backends/tests/nccl_two_rank.rs`, `crates/incin/src/train.rs`.

Depends on #98 and #82. Precedes #99.
